Cython/Utility/FunctionArguments.c

```diff
@@ -323,53 +424,93 @@ static int __Pyx_ParseOptionalKeywords(
 #endif
         Py_INCREF(value);
 
+        if (unlikely(!PyUnicode_Check(key))) goto invalid_keyword_type;
+
+        #if CYTHON_USE_UNICODE_INTERNALS
+        // The key hash is probably pre-calculated.
+        key_hash = ((PyASCIIObject*)key)->hash;
+        if (unlikely(key_hash == -1)) {
+            key_hash = PyObject_Hash(key);
+            if (unlikely(key_hash == -1))
+                goto bad;
+        }
+        #endif
+
+        // Compare strings for non-interned matches.
         name = first_kw_arg;
-        if (likely(PyUnicode_Check(key))) {
-            while (*name) {
-                int cmp = (
-                #if CYTHON_ASSUME_SAFE_SIZE
-                    (PyUnicode_GET_LENGTH(**name) != PyUnicode_GET_LENGTH(key)) ? 1 :
-                #endif
-                    PyUnicode_Compare(**name, key)
-                );
+        while (*name) {
+            PyObject *name_str = **name;
+            #if CYTHON_USE_UNICODE_INTERNALS
+            // Our argument hash is definitely pre-calculated.
+            if (key_hash == ((PyASCIIObject*)name_str)->hash)
+            #elif CYTHON_ASSUME_SAFE_SIZE
+            if (PyUnicode_GET_LENGTH(name_str) == PyUnicode_GET_LENGTH(key))
+            #endif
+            {
+                #if CYTHON_USE_UNICODE_INTERNALS
+                if (__Pyx_UnicodeKeywordsEqual(name_str, key))
+                #else
+                int cmp = PyUnicode_Compare(name_str, key);
                 if (cmp < 0 && unlikely(PyErr_Occurred())) goto bad;
-                if (cmp == 0) {
+                if (cmp == 0)
+                #endif
+                {
                     values[name-argnames] = value;
 #if CYTHON_AVOID_BORROWED_REFS
                     value = NULL;  /* ownership transferred to values */
 #endif
                     break;
                 }
-                name++;
             }
-            if (*name) continue;
-            else {
-                // not found after positional args, check for duplicate
-                PyObject*** argname = argnames;
+            name++;
+        }
+        if (*name) continue;
+
+        // Not found after positional args, check for (unlikely) duplicate positional argument.
```

Comment:
```suggestion
        // Not found after keyword args, check for (unlikely) duplicate positional argument.
```
I think?

Reply:
I'm not sure. I meant "the argument name was not found in the (keyword) argument names that follow the positional arguments". Clearly not a good comment.